src/ast/passes/semantic_analyser.cpp

```diff
@@ -2898,6 +2946,54 @@ void SemanticAnalyser::visit(AssignConfigVarStatement &assignment)
   Visit(assignment.expr);
 }
 
+void SemanticAnalyser::visit(VarDeclStatement &decl)
+{
+  const std::string &var_ident = decl.var->ident;
+
+  if (!IsValidVarDeclType(decl.var->type)) {
+    LOG(ERROR, decl.loc, err_)
+        << "Invalid variable declaration type: " << decl.var->type;
+  }
+
+  // This should be the first time we're seeing this variable
+  if (auto decl_search = variable_decls_[scope_].find(var_ident);
+      decl_search != variable_decls_[scope_].end()) {
+    if (decl_search->second != decl.loc) {
+      // Note this will change with proper block/lexical scoping
```

Comment:
Will be confusing if we never end up adding new scoping rules
```suggestion
      // Note this will change if/when proper block/lexical scoping is added to language
```